packages/react-tooltip/src/components/Tooltip/useTooltip.tsx

```diff
@@ -30,52 +45,117 @@ export const useTooltip = (
   ref: React.Ref<HTMLElement>,
   defaultProps?: TooltipProps,
 ): TooltipState => {
-  const [target, setTarget] = React.useState<HTMLElement | null>(null);
+  const [visible, setVisible] = React.useState(false);
 
   const state = mergeProps(
     {
       ref,
-      as: 'div',
+      children: props.children,
+      content: {
+        as: React.Fragment,
+      },
+      id: useId('tooltip-'),
       role: 'tooltip',
       position: 'above',
       align: 'center',
       offset: 4,
-      arrow: { as: 'div', children: '' },
-      visible: !!target,
+      showDelay: 250,
+      hideDelay: 250,
+      visible,
+      rendered: !useIsSSR(),
     },
     defaultProps && resolveShorthandProps(defaultProps, tooltipShorthandProps),
     resolveShorthandProps(props, tooltipShorthandProps),
   );
 
-  if (state.noArrow) {
-    state.arrow = undefined;
-  }
-
-  React.useImperativeHandle(
-    state.componentRef,
-    () => ({
-      show: t => setTarget(t),
-      hide: () => setTarget(null),
-    }),
-    [],
-  );
-
   const theme = useTheme();
 
-  const { targetRef, containerRef, arrowRef } = usePopper({
+  const popper = usePopper({
     enabled: state.visible,
     position: state.position,
     align: state.align,
     offset: [0, state.offset + (state.noArrow ? 0 : arrowHeight)],
     arrowPadding: theme?.global ? 2 * parseInt(tooltipBorderRadius(theme), 10) : 0,
   });
 
-  targetRef.current = target;
-  state.ref = useMergedRefs(state.ref, containerRef);
+  state.ref = useMergedRefs(state.ref, popper.containerRef);
+  state.arrowRef = popper.arrowRef;
+
+  const tooltipManager = useTooltipManager(createTooltipManager);
+
+  // Notify the manager when the pointer enters or leaves the tooltip
+  state.onPointerEnter = mergeCallbacks(() => tooltipManager.notifyEnterTooltip(), state.onPointerEnter);
```

Comment:
I think the way the event handlers are managed will stop users from optimizing the render of their own components when they place them as children of `Tooltip`, through `React.memo` or `React.PureComponent`

* `() => tooltipManager.notifyEnterTooltip` will create a new function reference for every render
* `mergeCallbacks` returns a new function reference for every time it's called
* If the users' own event handlers are not memoized then the same problem also happens

[This codesandbox](https://codesandbox.io/s/wild-cookies-j8cl9?file=/src/App.js) tries to show why `useEventCallback` hooks is used in our repo

Reply:
Great point. I reworked this to memoize the event listeners that are added to the child element.